impl/src/settings.rs

```diff
@@ -148,6 +149,63 @@ pub struct RazeSettings {
    */
   #[serde(default = "default_raze_settings_rust_rules_workspace_name")]
   pub rust_rules_workspace_name: String,
+
+  /**
+   * The expected path relative to the `Cargo.toml` file where vendored sources can
+   * be found. This should match the path passed to the `cargo vendor` command. eg:
+   * `cargo vendor -q --versioned-dirs "cargo/vendor"
+   */
+  #[serde(default = "default_raze_settings_vendor_dir")]
+  pub vendor_dir: String,
+}
+
+/** 
+ * Additional build files (see [crate::settings::CrateSettings::additional_build_file])
+ * are both representable by a path to a file and a name. This enum allows for this attribute
+ * to be representable by the path provided in the metadata (which is a relative path to a
+ * file from a `Cargo.toml` file) but resolvable to a canonicalized path so that the additional
+ * BUILD file may be accessed at any point. Build files that have relitive paths associated with
+ * them are considered `Validated`.
+ */
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(untagged)]
+pub enum AdditionalBuildFile {
+  Raw(String),
+  Validated { name: String, path: PathBuf },
+}
```

Comment:
This seems really clunky. IIUC, you're using a validation function to convert an initial "raw" string value into this "validated" value. You're masking over that with a bunch of From<> implementations.

Can we just parse this as a PathBuf and validate that it exists in validate_settings, instead of jumping through all of these hoops?

Reply:
The difficulty here is that if we have multiple packages that specify these files, we no longer know which package they came from once the settings are loaded since the `CrateSettings` are flattened into a single map. Additionally, if we simply use a `PathBuf` we end up writing canonicalized paths to files which is not ideal since it's not reproducible from one user to another when working on the same project. I then wondered if I can use the cargo workspace root as the anchor, but felt that the traceability/debugability was lacking in that users wouldn't be able to `grep` for any path rendered in a BUILD file through their `Cargo.toml` files to find problematic crate settings. Maybe that last option isn't actually so bad? My goal with this implementation was to try to keep this enum as invisible as possible.

With all that said, any new thoughts?

Reply:
> The difficulty here is that if we have multiple packages that specify these files, we no longer know which package they came from once the settings are loaded since the CrateSettings are flattened into a single map. 

What does this mean exactly? Where are the settings "flattened into a single map"?

I looked at create_crate_context, which is responsible from plumbing the settings from the settings object into the crate context object (which is later used below for cat-ing the contents of the "additional build file" into the generated file. This relies on `crate_settings`, which seems perfectly capable of distinguishing between different versions of the same crate.

> Additionally, if we simply use a PathBuf we end up writing canonicalized paths to files which is not ideal since it's not reproducible from one user to another when working on the same project. 

I don't know what this means either. What I drew from this was that there was some possibility of users specifying a fully qualified path here. I don't really know much about this feature because I've not been doing a good job reviewing all of this new stuff. Here's an example of an existing file:

https://github.com/google/cargo-raze/blob/06eaef173bf81f9b43f088a4243a8b7bd534c662/examples/remote/complicated_cargo_library/Cargo.toml#L32

That's not fully qualified.

What's wrong with parsing this as a pathbuf exactly?

> I then wondered if I can use the cargo workspace root as the anchor, but felt that the traceability/debugability was lacking in that users wouldn't be able to grep for any path rendered in a BUILD file through their Cargo.toml files to find problematic crate settings. Maybe that last option isn't actually so bad? 

I'm guessing that "using the cargo workspace as the anchor" means use the status quo situation where users specify a non-fully qualified path. That sounds right to me. 

> My goal with this implementation was to try to keep this enum as invisible as possible.

This enum is anything but invisible [for people who have to read and understand this code]

Reply:
> What does this mean exactly? Where are the settings "flattened into a single map"?

Oh, that actually relates to https://github.com/google/cargo-raze/pull/302 but looking at [settings.rs#L97](https://github.com/UebelAndre/cargo-raze/blob/manifest-path/impl/src/settings.rs#L97), that map stores all crate settings and in the context of cargo workspaces, the assumption that a single crate setting's `additional_build_file` setting being relative to the cargo workspace root is going to be incorrect in almost every single case. The crate settings in that PR get collected from each crate and stored in that map. If there was no workspace support, then yeah, I think it'd be easy to parse this as a `PathBuf` but I don't think that works here.

> I don't know what this means either. What I drew from this was that there was some possibility of users specifying a fully qualified path here. I don't really know much about this feature because I've not been doing a good job reviewing all of this new stuff. Here's an example of an existing file:

I'd initially tried parsing `additional_build_file` as a `PathBuf` which kinda but again, if these paths were to be canonicalized (which they need to be especially after https://github.com/google/cargo-raze/pull/302) then the full path (including potential user information) would be written by [bazel.rs#L228-L231](https://github.com/UebelAndre/cargo-raze/blob/manifest-path/impl/src/rendering/bazel.rs#L228-L231). So the paths written there need to at least be relative to the cargo workspace root.

> I'm guessing that "using the cargo workspace as the anchor" means use the status quo situation where users specify a non-fully qualified path. That sounds right to me.

I'll see if there's a solution here. There likely is.

> This enum is anything but invisible [for people who have to read and understand this code]

I need to read your next message, but I meant to say "invisible for users, not necessarly contributors. I felt these ~40 lines of code were fairly straight forward so I don't think they're quite magic. But again, I'll read https://github.com/google/cargo-raze/pull/301#issuecomment-741495411 and reply.